### PR TITLE
BUG: check return value from PyArray_PromoteTypes

### LINKED ISSUE #12031: Segmentation fault PyArray_ConvertToCommonType.(NULL pointer derference)
NULL pointer derference in PyArray_ConvertToCommonType
```
import numpy
numpy.half(-1).choose(numpy.void(1), numpy.broadcast_arrays((1,)))
```
gdb info
```
[----------------------------------registers-----------------------------------]
RAX: 0xc0400001756 --> 0x0 
RBX: 0x0 
RCX: 0x0 
RDX: 0xffffdad7614 --> 0x0 
RSI: 0x2 
RDI: 0xc03700 --> 0x24c3 
RBP: 0x7fffed6bb0a0 --> 0x3 
RSP: 0x7fffffffcf30 --> 0xffffffff9ec --> 0x0 
RIP: 0x7ffff2803769 (<PyArray_ConvertToCommonType+1865>:	add    QWORD PTR [rbx],0x1)
R8 : 0x501 
R9 : 0x7ffff3ac653d --> 0xfff7eafeb000006e 
R10: 0x5 
R11: 0x7fffffffce30 --> 0xbe90a0 --> 0xb ('\x0b')
R12: 0x0 
R13: 0x60200000bab0 --> 0x0 
R14: 0x0 
R15: 0x0
EFLAGS: 0x10246 (carry PARITY adjust ZERO sign trap INTERRUPT direction overflow)
[-------------------------------------code-------------------------------------]
   0x7ffff2803755 <PyArray_ConvertToCommonType+1845>:	je     0x7ffff28039f0 <PyArray_ConvertToCommonType+2512>
   0x7ffff280375b <PyArray_ConvertToCommonType+1851>:	cmp    BYTE PTR [r15+0x7fff8000],0x0
   0x7ffff2803763 <PyArray_ConvertToCommonType+1859>:	jne    0x7ffff2804188 <PyArray_ConvertToCommonType+4456>
=> 0x7ffff2803769 <PyArray_ConvertToCommonType+1865>:	add    QWORD PTR [rbx],0x1
   0x7ffff280376d <PyArray_ConvertToCommonType+1869>:	xor    ecx,ecx
   0x7ffff280376f <PyArray_ConvertToCommonType+1871>:	xor    r9d,r9d
   0x7ffff2803772 <PyArray_ConvertToCommonType+1874>:	xor    edx,edx
   0x7ffff2803774 <PyArray_ConvertToCommonType+1876>:	mov    rsi,rbx
[------------------------------------stack-------------------------------------]
0000| 0x7fffffffcf30 --> 0xffffffff9ec --> 0x0 
0008| 0x7fffffffcf38 --> 0x7fffffffd040 --> 0x7fff00000002 
0016| 0x7fffffffcf40 --> 0x7ffff3ace0b0 --> 0x1 
0024| 0x7fffffffcf48 --> 0x7fff00000005 
0032| 0x7fffffffcf50 --> 0x60200000bab0 --> 0x0 
0040| 0x7fffffffcf58 --> 0x0 
0048| 0x7fffffffcf60 --> 0x100000002 --> 0x0 
0056| 0x7fffffffcf68 --> 0x7fffed70f210 --> 0x2 
[------------------------------------------------------------------------------]
Legend: code, data, rodata, value
Stopped reason: SIGSEGV
0x00007ffff2803769 in PyArray_ConvertToCommonType (op=op@entry=0x7ffff3ace0b0, retn=retn@entry=0x7fffffffd040) at numpy/core/src/multiarray/convert_datatype.c:2135
2135	numpy/core/src/multiarray/convert_datatype.c: No such file or directory.
gdb-peda$ bt
#0  0x00007ffff2803769 in PyArray_ConvertToCommonType (op=op@entry=0x7ffff3ace0b0, retn=retn@entry=0x7fffffffd040) at numpy/core/src/multiarray/convert_datatype.c:2135
#1  0x00007ffff28d3ed3 in PyArray_Choose (ip=ip@entry=0x7fffedb85c10, op=0x7ffff3ace0b0, out=0x0, clipmode=<optimized out>) at numpy/core/src/multiarray/item_selection.c:671
#2  0x00007ffff294f51d in array_choose (self=<optimized out>, args=0x7ffff3ace0b0, kwds=0x0) at numpy/core/src/multiarray/methods.c:1124
#3  0x000000000046da0e in cfunction_call_varargs (kwargs=<optimized out>, args=<optimized out>, func=0x7fffedb90380) at Objects/call.c:753
#4  PyCFunction_Call (func=0x7fffedb90380, args=<optimized out>, kwargs=<optimized out>) at Objects/call.c:784
#5  0x00007ffff29c333b in gentype_generic_method (self=<optimized out>, args=0x7ffff3ace0b0, kwds=0x0, str=<optimized out>) at numpy/core/src/multiarray/scalartypes.c.src:187
#6  0x000000000046b310 in _PyMethodDef_RawFastCallKeywords (method=0x7ffff2cffde0 <gentype_methods+832>, self=self@entry=0x7ffff00cec78, args=args@entry=0x7ffff7eb11b8, nargs=nargs@entry=0x2, 
    kwnames=kwnames@entry=0x0) at Objects/call.c:690
#7  0x000000000081403d in _PyMethodDescr_FastCallKeywords (descrobj=descrobj@entry=0x7ffff3a452c0, args=args@entry=0x7ffff7eb11b0, nargs=nargs@entry=0x3, kwnames=kwnames@entry=0x0)
    at Objects/descrobject.c:288
#8  0x00000000004416b2 in call_function (kwnames=0x0, oparg=0x3, pp_stack=<synthetic pointer>) at Python/ceval.c:4600
#9  _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:3186
#10 0x00000000005ecfbc in PyEval_EvalFrameEx (throwflag=0x0, f=0x7ffff7eb1040) at Python/ceval.c:536
#11 _PyEval_EvalCodeWithName (_co=_co@entry=0x7ffff7eb7b70, globals=globals@entry=0x7ffff7eb11a8, locals=locals@entry=0x7ffff7eb7bc8, args=args@entry=0x0, argcount=argcount@entry=0x0, 
    kwnames=kwnames@entry=0x0, kwargs=0x0, kwcount=0x0, kwstep=0x2, defs=0x0, defcount=0x0, kwdefs=0x0, closure=0x0, name=0x0, qualname=0x0) at Python/ceval.c:3951
#12 0x00000000005ed4d3 in PyEval_EvalCodeEx (closure=0x0, kwdefs=0x0, defcount=0x0, defs=0x0, kwcount=0x0, kws=0x0, argcount=0x0, args=0x0, locals=locals@entry=0x7ffff7eb7bc8, 
    globals=globals@entry=0x7ffff7eb11a8, _co=_co@entry=0x7ffff7eb7b70) at Python/ceval.c:3980
#13 PyEval_EvalCode (co=co@entry=0x7ffff7eb7b70, globals=globals@entry=0x7ffff7ef6080, locals=locals@entry=0x7ffff7ef6080) at Python/ceval.c:513
#14 0x0000000000687cdd in run_mod (arena=0x7ffff7f72078, flags=0x7fffffffda60, locals=0x7ffff7ef6080, globals=0x7ffff7ef6080, filename=0x7ffff7e6cb90, mod=0x625000020ce0) at Python/pythonrun.c:1031
#15 PyRun_FileExFlags (fp=<optimized out>, filename_str=<optimized out>, start=<optimized out>, globals=<optimized out>, locals=<optimized out>, closeit=<optimized out>, flags=0x7fffffffda60)
    at Python/pythonrun.c:984
#16 0x000000000068817b in PyRun_SimpleFileExFlags (fp=<optimized out>, filename=0x7ffff7e88f50 "poc2.py", closeit=0x1, flags=<optimized out>) at Python/pythonrun.c:425
#17 0x000000000044d9a4 in pymain_run_file (config=0x61e00000f108, cf=0x7fffffffda60, pymain=0x7fffffffdc60) at Modules/main.c:1159
#18 pymain_run_python (interp=0x61e00000f080, pymain=0x7fffffffdc60) at Modules/main.c:1607
#19 pymain_main (pymain=pymain@entry=0x7fffffffdc60) at Modules/main.c:1755
#20 0x000000000044e39c in _Py_UnixMain (argc=0x2, argv=0x7fffffffdde8) at Modules/main.c:1792
#21 0x00007ffff6193830 in __libc_start_main (main=0x424060 <main>, argc=0x2, argv=0x7fffffffdde8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffddd8)
    at ../csu/libc-start.c:291
#22 0x0000000000446229 in _start ()
```

numpy.version.version
'1.15.2'

Python 3.8.0a0 (heads/master:f6c8007, Sep 25 2018, 12:42:29) 
[GCC 5.4.0 20160609] on linux

### PR DESCRIPTION
**Pull request by mattip**
Fixes #12031. by checking the return value from `PyArray_PromoteTypes`. Also make sure to decref `otmp` on error.

### CODE REVIEW

**Comment by charris** [comment]
I have no idea why the debug build is segfaulting.

**Review by eric-wieser — CHANGES_REQUESTED** [review]
This isn't safe. There are two loops here - one of the decrefs no matter how it exits, and the other doesn't. I think the thing to do here is push `otmp` into the local scope of each loop.

**Comment by mattip** [comment]
Fixing `otmp`, the decref caused the segfault.

File: numpy/core/src/multiarray/convert_datatype.c [hunk 1]
```diff
@@ -2067,11 +2066,15 @@ PyArray_ConvertToCommonType(PyObject *op, int *retn)
...
 
     for (i = 0; i < n; i++) {
-        otmp = PySequence_GetItem(op, i);
+        PyObject *otmp = PySequence_GetItem(op, i);
+        if (otmp == NULL) {
+            goto fail;
+        }
         if (!PyArray_CheckAnyScalar(otmp)) {
             newtype = PyArray_DescrFromObject(otmp, intype);
             Py_XDECREF(intype);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/convert_datatype.c]: Put the decref on this line, and then you can omit it below.

File: numpy/core/src/multiarray/convert_datatype.c [hunk 2]
```diff
@@ -2081,14 +2084,15 @@ PyArray_ConvertToCommonType(PyObject *op, int *retn)
             newtype = PyArray_DescrFromObject(otmp, stype);
             Py_XDECREF(stype);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/convert_datatype.c]: Same again - you're done with the object already by this line, so can decref it immediately

Comment (mattip) [reply numpy/core/src/multiarray/convert_datatype.c]: done

File: numpy/core/src/multiarray/convert_datatype.c (hunk 1 above)
Comment (mattip) [reply numpy/core/src/multiarray/convert_datatype.c]: done

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Thanks @mattip!

**Pull request merged**